Fix conflict by disabling Yoast initialization on Divi preview

renatho opened the pull request:
Fixes temporarily #6112

### Changes proposed in this Pull Request

* In order to fix the conflict between Learning Mode + Divi + Yoast, this PR disables the Yoast initialization when it's running the Divi preview in the editor.
* It's important to mention that I didn't find the root cause of the issue yet, but since this solution is probably safe I decided to stop investing much more time in that. I don't think a user would want to optimize the SEO in the Divi preview in the editor - unless there's any other implication that I didn't realize.

### My findings from the investigation

* `\add_action( 'wpseo_head', [ $this, 'present_head' ], -9999 );` in Yoast will make it call the `get_the_excerpt` (`Post_Helper::get_the_excerpt`). If we don't call the `get_the_excerpt` there, the problem doesn't happen.
* There's also a method in Divi that I suspect a little, but I couldn't figure out too much from that. The method is `ET_Builder_Element::reset_element_indexes`. It seems it tries to fix something related to `the_content` recursions. And my suspicious is that it has something related to the problem because with the Learning Mode we make some nested `the_content` calls in `Sensei\Blocks\Course_Theme\Course_Content::render_content`.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* See the [issue details](https://github.com/Automattic/sensei/issues/6112), and make sure you can reproduce it using the `trunk`.
* Check that running the code from this branch the problem doesn't happen anymore.

renatho committed:
Disable Yoast initialization on Divi preview

renatho commented on includes/3rd-party/divi.php:
```diff
@@ -0,0 +1,17 @@
+ * It avoids an existing conflict that prevents the preview from loading.
+ */
+function sensei_fix_divi_yoast_conflict() {
+	if ( isset( $_GET['et_block_layout_preview'] ) && '1' === $_GET['et_block_layout_preview'] ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
The ideal would be to also have a check to run it only in lessons with Learning Mode, but it's too early to check this.

jom reviewed (dismissed):
This fixes the described issue. I wonder if after this is merged we should transfer your notes to a new issue. It makes sense to me to merge this now, but we should probably prioritize finding a less blunt solution.

jom reviewed (commented):
Actually, I just had a thought. This allows non-authenticated users to prevent loading Yoast from any request. It seems a bit too powerful. Is there a way we can restrict it just to the required request?

renatho committed:
Add extra check to make sure the request is coming from the admin

renatho commented:
> Actually, I just had a thought. This allows non-authenticated users to prevent loading Yoast from any request. It seems a bit too powerful. Is there a way we can restrict it just to the required request?

Hi Jake! Thank you for your review! That's a great point! Since it's in the `plugins_loaded` hook, we don't have too much information to check, but I think the extra check I added is enough: https://github.com/Automattic/sensei/pull/6342/commits/eea3a403314700defcbcdc34d7d919bcb2b52caf. It will check if the request is coming from the admin. WDYT?

> we should probably prioritize finding a less blunt solution.

Totally agree! For now, I just added [this comment](https://github.com/Automattic/sensei/issues/6112#issuecomment-1372545691) to the original issue, changing it to low priority. So we can keep all the history of this issue in the same issue. Does it make sense?

renatho committed:
Refactor solution to make it safer

renatho commented:
@jom, I checked that we could check the user roles (`edit_lessons`, for example) in the `plugins_loaded`, but based on your idea, I sent another commit here changing the hook and removing all actions from `wpseo_head`: https://github.com/Automattic/sensei/pull/6342/commits/932fc836dcd2a7ba5c1a89282428a9a63107d386. So we check if they can edit specifically that post.

**Update:** I also added another check to remove the action only when it's a lesson with learning mode enabled. https://github.com/Automattic/sensei/pull/6342/commits/86d4422ec55e7a6e0d9c6378e2e04b6bb1fe43e4

renatho committed:
Add extra check to remove action only on lessons with learning mode enabled

jom reviewed (approved):
Looks great! Thanks for picking this one up. 🪂
